torchtext/datasets/language_modeling.py

```diff
@@ -1,217 +1,236 @@
+            >>> data = torch.tensor([token_id_1, token_id_2,
+                                     token_id_3, token_id_1]).long()
+            >>> vocab = build_vocab_from_iterator([['language', 'modeling']])
+            >>> dataset = LanguageModelingDataset(data, vocab)
 
-        This is the simplest way to use the dataset, and assumes common
-        defaults for field, vocabulary, and iterator parameters.
-
-        Arguments:
-            batch_size: Batch size.
-            bptt_len: Length of sequences for backpropagation through time.
-            device: Device to create batches on. Use -1 for CPU and None for
-                the currently active GPU device.
-            root: The root directory that the dataset's zip archive will be
-                expanded into; therefore the directory in whose wikitext-2
-                subdirectory the data files will be stored.
-            wv_dir, wv_type, wv_dim: Passed to the Vocab constructor for the
-                text field. The word vectors are accessible as
-                train.dataset.fields['text'].vocab.vectors.
-            Remaining keyword arguments: Passed to the splits method.
         """
-        TEXT = data.Field()
-
-        train, val, test = cls.splits(TEXT, root=root, **kwargs)
-
-        TEXT.build_vocab(train, vectors=vectors)
-
-        return data.BPTTIterator.splits(
-            (train, val, test), batch_size=batch_size, bptt_len=bptt_len,
-            device=device)
 
+        super(LanguageModelingDataset, self).__init__()
+        self._data = data
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _get_datafile_path(key, extracted_files):
+    for fname in extracted_files:
+        if key in fname:
+            return fname
+
+
+def _setup_datasets(dataset_name, tokenizer=get_tokenizer("basic_english"),
+                    root='.data', vocab=None, removed_tokens=[],
+                    data_select=('train', 'test', 'valid')):
+
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    for item in data_select:
+        if item not in ('train', 'test', 'valid'):
+            raise TypeError('{} in data_select is not supported!'.format(item))
+
+    if dataset_name == 'PennTreebank':
+        extracted_files = []
+        if 'train' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][0], root=root))
+        if 'test' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][1], root=root))
+        if 'valid' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][2], root=root))
+    else:
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
```

Comment:
Another thing we could do with these download functions is compare the result to a checksum. We added those capabilities in torchaudio. I think that might actually be worth it, if this checksum calculating is fast. If it's not fast we might need to make this a flag. This could save a lot of grief and is something rather common. Should we do this now? cc @vincentqb

Reply:
Having a checksum is a good idea. However, since the goal is to soonish put the download function in sync, we should avoid investing time optimizing this particular version. I'd recommend using the one we worked on recently in [torchaudio](https://github.com/pytorch/audio/blob/master/torchaudio/datasets/utils.py#L106) instead.

Reply:
Yes we should copy this one into here as well. Do you want to create a PR?

Reply:
@vincentqb @cpuhrsch I'm fine to put one `download` function across domains if it doesn't break any current download activity.

Reply:
I think it's worth testing this, since it's our goal down the road :)